tests/fixtures/expanders.php

```diff
@@ -13,3 +13,63 @@ function my_hello_expander(TokenStream $ts) : TokenStream {
         )
     ;
 }
+
+function reverse_ast_expander(ParsedTokenStream $stream, Engine $engine) : ParsedTokenStream {
+    $ast = $stream->getAst();
+
+    $tokens = $ast->tokens();
+    $first = array_shift($tokens);
+
+    _modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
+        $value["outer"]["inner"][0] = new Token($first->type(), strrev($first->value()));
+        return $value;
+    });
+
+    return ParsedTokenStream::fromSource(
+        join(" ", $ast->tokens())
+    );
+}
+
+function upper_ast_expander(ParsedTokenStream $stream, Engine $engine) : ParsedTokenStream {
+    $ast = $stream->getAst();
+
+    $tokens = $ast->tokens();
+    $first = array_shift($tokens);
+
+    _modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
+        $value["outer"]["inner"][0] = new Token($first->type(), strtoupper($first->value()));
+        return $value;
+    });
+
+    return ParsedTokenStream::fromSource(
+        join(" ",  $ast->tokens())
+    );
+}
+
+function wrap_ast_expander(ParsedTokenStream $stream, Engine $engine) : ParsedTokenStream {
+    $ast = $stream->getAst();
+
+    $tokens = $ast->tokens();
+    $first = array_shift($tokens);
+
+    _modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
+        $value["outer"]["inner"][0] = new Token($first->type(), "[" . $first->value() . "]");
+        return $value;
+    });
```

Comment:
```diff
-$tokens = $ast->tokens();
-$first = array_shift($tokens);
-
-$tokens = $ast->tokens();
-$first = array_shift($tokens);
-_modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
-    $value["outer"]["inner"][0] = new Token($first->type(), "[" . $first->value() . "]");
-    return $value;
-});
+$leaf = $ast->{'* outer inner 0'}->token();
+new Ast($ast->label(), ['outer' => ['inner' => new Token($leaf->type(), "[{$leaf}]")]]);
```

Reply:
It's not enough just to return the new AST, it needs to be modified in-place before the next expander happens.

Reply:
Having tried that approach, I do not think it'll work. Creating the new AST changes the labels of the AST structure so that subsequent accesses cannot resolve the same data points.

Reply:
Mostly because it's not good for people to see tricky code at tests. The less the better. At some point I had either the expectation of seeing:

```php
$cg->context = $mutation->getAst(); // on the Engine.php, after the expansion
```

Or if, we don't want to give that control:

```diff
-$tokens = $ast->tokens();
-$first = array_shift($tokens);
-
-$tokens = $ast->tokens();
-$first = array_shift($tokens);
-_modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
-    $value["outer"]["inner"][0] = new Token($first->type(), "[" . $first->value() . "]");
-    return $value;
-});
+$token = $ast->{'* outer inner 0'}->token();
+$ast->{'outer inner 0'} = new Token($token->type(), "[{$leaf}]", $token->line());
```

To support that latest suggestion, and improve the Ast situation we have been struggling with, I've just added this at master 58556ccc676740f1e4397d546e192d032726e878.

Reply:
That's much better! Updated.